### PR TITLE
[crossgen2] Use Alignment to calculate class Offset

### PR DESCRIPTION
**Pull request by t-mustafin**
`ComputeAutoFieldLayout()` function ignores calculated `fieldSizeAndAlignment.Alignment` and uses `context.Target.LayoutPointerSize` instead. It produces a problems if `fieldSizeAndAlignment.Alignment > context.Target.LayoutPointerSize` on arm for example.
I suggest to use maximum of that values to calculate offset.
Issue #38822

### CODE REVIEW

**Comment by t-mustafin** [comment]
cc @alpencolt @jkotas @mangod9 @davidwrighton

**Review by pavel-orekhov — CHANGES_REQUESTED** [review]

File: src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs [hunk 1]
```diff
@@ -633,7 +633,8 @@ protected ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type,
                 }
                 else
                 {
```
[the remark this answers is not in the document]
Comment (t-mustafin) [reply src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:635]: Simply invert condition works in my case, but [comment](https://github.com/dotnet/runtime/blob/46600ec2a4b966d51fc25f6960db3ce114ef5246/src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs#L625) says that the condition is matched correctly.

**Comment by t-mustafin** [comment]
cc @MichalStrehovsky

**Comment by davidwrighton** [comment]
I've written a more comprehensive layout test (See https://github.com/davidwrighton/runtime/tree/layout_testing ) That test isn't ready to merge, as the amount of C# in the test is unreasonably large, but I believe it should be a good way to see that most of these cases are handled properly. I plan to run that test on arm32 tomorrow, or Monday, with this fix.

File: src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs [hunk 2]
```diff
@@ -633,7 +633,8 @@ protected ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type,
...
-                    cumulativeInstanceFieldPos = LayoutInt.AlignUp(cumulativeInstanceFieldPos, context.Target.LayoutPointerSize, context.Target);
+                    largestAlignmentRequired = LayoutInt.Max(fieldSizeAndAlignment.Alignment, context.Target.LayoutPointerSize);
```
Comment (davidwrighton) [inline src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:636]: This value doesn't represent the largetAlignment required, so please don't re-use the variable here. Define a new variable.

**Review by davidwrighton — COMMENTED** [review]
I was unable to run the layout test to completion on the arm32 hardware I have available. I repeatedly saw crashes while attempting to run the JIT_GetRuntimeType helper. However, my analysis of the situation indicates that @MichalStrehovsky is correct, handling of the FEATURE_64BIT_ALIGNMENT behavior in crossgen2 is what needs to be matched. The particular fix here looks to be correct, and appears to match up with this snippet of the type loader in coreclr.

https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/vm/methodtablebuilder.cpp#L8113-L8119

My analysis of the rest of the behavior of the runtime indicates that this is at least close to correct, but as I was unable to run the test suites, I could not confirm further details. If you fix the vairable use issue I mentioned here, I will approve this fix as it is strictly more correct than the previous state, but I do not have confidence that all issues have been found.

File: src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs (hunk 2 above)
Comment (t-mustafin) [reply src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: Thanks, I defined new variable `AlignmentRequired`.

**Review by davidwrighton — APPROVED** [review]
As I said earlier, I'm not entirely convinced this fixes everything, but I think its a strict improvement over the current state.

**Pull request merged**